lib/charms/vault_k8s/v0/vault_s3.py

```
@@ -0,0 +1,217 @@
+from mypy_boto3_s3.service_resource import Bucket
+from mypy_boto3_s3.type_defs import CreateBucketConfigurationTypeDef
+
+# The unique Charmhub library identifier, never change it
+LIBID = "6a14cfe8d3134db4a6c47c4cf7b7d5d6"
+
+# Increment this major API version when introducing breaking changes
+LIBAPI = 0
+
+# Increment this PATCH version before using `charmcraft publish-lib` or reset
+# to 0 if you are raising the major API version
+LIBPATCH = 2
+
+
+class LogAdapter(logging.LoggerAdapter):
+    """Adapter for the logger to prepend a prefix to all log lines."""
+
+    prefix = "vault_s3"
+
+    def process(self, msg, kwargs):
+        """Decides the format for the prepended text."""
+        return f"[{self.prefix}] {msg}", kwargs
+
+
+logger = LogAdapter(logging.getLogger(__name__), {})
+
+AWS_DEFAULT_REGION = "us-east-1"
+
+
+class S3Error(Exception):
+    """Base class for S3 errors."""
+
+    pass
+
+
+class S3:
+    """A class representing an S3 session allowing S3 operations."""
+
+    def __init__(
+        self,
+        access_key: str,
+        secret_key: str,
+        endpoint: str,
+        region: str | None = AWS_DEFAULT_REGION,
+    ):
+        self.access_key = access_key
+        self.secret_key = secret_key
+        self.endpoint = endpoint
+        self.region = region
+        try:
+            self.session = boto3.session.Session(
+                aws_access_key_id=self.access_key,
+                aws_secret_access_key=self.secret_key,
+                region_name=region,
+            )
+            custom_config = Config(
+                retries={
+                    "max_attempts": 1,
+                },
+            )
+            self.s3 = self.session.resource("s3", endpoint_url=self.endpoint, config=custom_config)
+        except (ClientError, BotoCoreError, ValueError) as e:
+            raise S3Error(f"Error creating session: {e}")
+
+    def create_bucket(self, bucket_name: str) -> bool:
```

I'm not sure this method and some others in this lib actually provide significant value. They seem to just wrap boto code and mask exceptions in favour of return values, limiting the developer's options

---

Yes correct, but this lib is meant for internal use in the Vault charms.
We have such libs because we maintain both the k8s and the machine charms and we wanted a simple way of sharing code between both charms

---

Understood. Thanks for clarifying